src/plugins/snapshot/snapshot-listener.ts

```diff
@@ -44,61 +55,65 @@ export class SnapshotListener implements GraphQLRequestListener<PolarisGraphQLCo
-                    method: httpRequest.method,
-                    request: httpRequest,
-                    query: requestContext.request,
-                    options: {
-                        ...SnapshotListener.graphQLOptions,
+            if (!context.snapshotContext) {
+                const totalCount = firstRequest.extensions.totalCount;
+                if (totalCount != null) {
+                    this.fillContextWithSnapshotMetadata(
                         context,
-                    },
-                });
-
-                const parsedResult = JSON.parse(currentPageResult.graphqlResponse);
-
-                if (!context.snapshotContext) {
-                    const totalCount = parsedResult.extensions.totalCount;
-                    if (totalCount !== undefined) {
-                        context.snapshotContext = {
-                            totalCount,
-                            startIndex: 0,
-                            countPerPage: requestHeaders.snapPageSize
-                                ? Math.min(
-                                      this.snapshotConfiguration.maxPageSize,
-                                      requestHeaders.snapPageSize,
-                                  )
-                                : this.snapshotConfiguration.maxPageSize,
-                        };
-                        context.returnedExtensions.globalDataVersion =
-                            parsedResult.extensions.globalDataVersion;
-                    } else {
-                        return;
-                    }
+                        totalCount,
+                        requestHeaders,
+                        firstRequest,
+                    );
+                } else {
+                    return;
                 }
+            }
 
-                context.snapshotContext!.prefetchBuffer = parsedResult.extensions.prefetchBuffer;
-                delete parsedResult.extensions.prefetchBuffer;
-                const snapshotPage = new SnapshotPage(JSON.stringify(parsedResult));
-                await snapshotRepository.save({} as any, snapshotPage);
-                pagesIds.push(snapshotPage.getId());
-                context.snapshotContext!.startIndex! += context.snapshotContext!.countPerPage!;
-                currentPageIndex++;
-            } while (
-                currentPageIndex <
-                context.snapshotContext!.totalCount! / context.snapshotContext!.countPerPage!
+            context = cloneDeep(context);
+            const pageCount = Math.ceil(
+                context.snapshotContext!.totalCount! / context.snapshotContext!.countPerPage!,
             );
-
-            context.returnedExtensions.snapResponse = { pagesIds };
+            const snapshotPages: SnapshotPage[] = [];
+            const pagesIds: string[] = Array(pageCount)
+                .fill(0)
+                .map(function generateUUIDAndCreateSnapshotPage() {
```

Comment:
Declare this function outside of the current function, and tranform into an arrow function